### DISCUSSION TITLE
Questions regarding DIA-NN protein inference

### DISCUSSION THREAD

**Discussion by hguturu** [vdemichev/DiaNN#316] [Q&A]
For this discussion, we have two main modes of running DIA-NN one with a MSFragger based DDA library (`.tsv` file) and another is in silico generated library (`.speclib` file i.e. Library Free). 
In trying to understand protein inference, we came across what we believe are 3 relevant flags 
- `--relaxed-prof-inf` 
- `--pg-level {0/1/2}` 
- `--reannotate` 
# `--relaxed-prot-inf` 
Our understanding is this changes DIA-NN default protein inference algorithm from creating new protein groups for razor peptides, by heuristically assigning the peptides to a single group. But, both this and the default protein inference do greedy set cover. Is this correct?
e.g. if peptide A maps to Protein 1, peptide B maps to Protein 1 and Protein 2 and peptide C maps to Protein 3 
- Default inference would create 3 protein groups  
  - Protein 1 
  - Protein 1;Protein2 
  - Protein 2 
- `--relaxed-prot-inf` would only create 2 protein groups 
  - Protein 1 (by some heuristic assign peptide B to Protein 1 only) 
  - Protein 2 

Questions:
- Do the entries in `Protein.Group` and `Protein.Ids` in `report.tsv` have different meaning with these flags? Or is `Protein.Ids` always the list of fasta records the peptide maps to and `Protein.Groups` is subset that is finally inferred? 
  - i.e. for Peptide B in above example is `Protein.Ids` always `Protein 1;Protein2`, but `Protein.Groups` changes based on which inference engine is used?
- This flag has no effect when using a DDA library unless it's paired with `--reannotate`, correct?
# `--pg-level {0/1/2}`
This flag changes the definition of proteotypic from a peptide uniquely mapping to an isoform (0) to protein (1) to gene (2).

Questions:
- When using a Uniprot search database are `--pg-level 0` and `--pg-level 1` equivalent since all FASTA records have a different accession ID? If not, what fields is DIA-NN parsing to generate these assignments
- Is using `--pg-level 1` recommended when benchmarking against other tools?
- For `--pg-level 2` is DIA-NN parsing the `GN=` field in the Uniprot header?
- Is it expected for `--pg-level 1` to yield slightly fewer protein groups due to not having any proteotypic peptides?

# `--reannotate`
Our initial understanding of this flag was it tells DIA-NN to turn on protein inference when using a DDA library.

Questions:
- How does it reannoate from Fasta header? Is the logic a particular peptide is mapped to all the sequences in the Fasta database and then those IDs are used? Will existing IDs be overwriten?
- Why does this change how protein inference is done for library-free mode when the .speclib file was previously generated from the same Fasta sequence?
- For benchmarking, what is the recommended combination of flags? Is it `--relaxed-prof-inf --pg-level 1 –reannotate` to be most in line with other tools?

**Comment by vdemichev** [comment]
--relaxed-prot-inf:
- Protein.Ids lists all sequence identifiers associated with the peptide.
- This flag has effect, unless protein inference is turned off.

--pg-level:
- I am not sure whether all sequences in uniprot have protein names associated with them.
- Please see the docs on what's recommended. The only reason to change --pg-level is either if you are interested in isoforms or if there's no proper gene annotation available.
- Yes.
- Not sure about this.

--reannotate:
- Existing IDs are overwritten. DIA-NN reannotates by identifying proteins from which the peptide could originate considering the specified digestion enzyme.
- I am not sure what you mean here.
- Please follow recommendation in the docs. The main principle: keep settings default, unless recommended otherwise for the specific scenario.

**Reply by hguturu** [reply]
>Why does this change how protein inference is done for library-free mode when the .speclib file was previously generated from the >same Fasta sequence?
>> I am not sure what you mean here.

When we run DIA-NN in library free mode using a `.speclib` file we get different protein grouping if we run it with and without `--reannotate`. We are not sure why, because we though `--reannotate` is only relevant when using a DDA library with pre-computed groups which the generated library wouldn't have.

That is if we try:
`/usr/diann/1.8/diann-1.8 --threads 8 --dir raw --fasta UP000005640_9606_combined.fasta --lib UP000005640_9606_combined.predicted.speclib --use-quant --mass-acc-ms1 10 --mass-acc 10 --qvalue 0.01 --matrices --missed-cleavages 1 --met-excision --cut K*,R* --smart-profiling --relaxed-prot-inf --pg-level 1`
and 
`/usr/diann/1.8/diann-1.8 --threads 8 --dir raw --fasta UP000005640_9606_combined.fasta --lib UP000005640_9606_combined.predicted.speclib --use-quant --mass-acc-ms1 10 --mass-acc 10 --qvalue 0.01 --matrices --missed-cleavages 1 --met-excision --cut K*,R* --smart-profiling --relaxed-prot-inf --pg-level 1 --reannotate`

We get different protein groups, where our understanding from the documentation is the predicted library shouldn't need reannotation like the DDA libraries? Note we are also using `--smart-profiling` above which is a recommended flag, but based on your other discussion response may not actually be doing anything?

>For benchmarking, what is the recommended combination of flags? Is it --relaxed-prof-inf --pg-level 1 –reannotate to be most in line with other tools?
>>Please follow recommendation in the docs. The main principle: keep settings default, unless recommended otherwise for the specific scenario.

Agree on using defaults, but the defaults seem to be written primarily for downstream analysis (as would make sense for most). But, we are curious on benchmarking the performance of this tool with other tools and alternate workflows (e.g. DIA vs DDA). Thus, we would you be able to recommend the flags that would be appropriate for most apples to apples comparison with other tools (e.g. MaxQuant / MSFragger for DDA and MaxDIA / MSFragger-DIA / EncyclopeDIA)? I understand you won't be able to recommend that without knowing the details of the other tools so ball park recommendations are file (e.g. your recommendation on using `--relaxed-prot-inf`). Thus we are aiming at `--relaxed-prot-inf --pg-level 1 --reannotate`, but are unsure if this is recommended for all comparisions (e.g. library free and DDA library) and if there are additional flags.

**Reply by hguturu** [reply]
>This flag has no effect when using a DDA library unless it's paired with --reannotate, correct?
>>This flag has effect, unless protein inference is turned off.
Thats curious since we didn't see any difference using a MSFragger library with or without `--relaxed-prot-inf` until we also used `--reannotate`. Is that expected?

**Reply by vdemichev** [reply]
Thank you for bringing this to my attention. What happens, is that regular digests considers peptide length limit (<= 30 by default) before methionine excision, while Reannotate does so after methionine excision and is thus less strict. I will make it consistent in the next release.

**Reply by vdemichev** [reply]
If unknown if proper gene annotation is present, OK to set DIA-NN's protein inference ot isoforms in combination with --relaxed-prot-inf for benchmarking.

**Comment by ht-lau** [comment]
I found that the protein inferencing in DIA-NN is quite different from other peptide search engines. I have an anti-CD19 CAR, fusion of the anti-CD19 scFv, a 4-1BB signaling domain, and the CD3z signal domain. I know that my samples does not have native 4-1BB and CD3z. In an experiment, 7 peptides were identified, 1 from CD19 scFv, 1 from 4-1BB, 5 from CD3z. Using default settings, DIA-NN quant matrix showed three separated rows for CD19 CAR, 4-1BB, CD3z. With --relaxed-prot-inf, it gave 2, CD19 CAR and CD3z. Using FragPipe, only CD19 CAR was output. I have not tested it on MaxQuant, but based on experience, MaxQuant would also roll them into one entity.

**Reply by vdemichev** [reply]
To each of DIA-NN's protein groups some peptides are mapped. So if two groups are identified, means there's evidence for both of them. FragPipe probably just identifies different peptides?

**Reply by vdemichev** [reply]
Naturally, if there are both CD19 peptides and CD3z peptides detected, both proteins should be reported as separate protein groups? So no surprise here?

**Reply by ht-lau** [reply]
Hi Vadim, 
The protein is an anti-CD19 chimeric receptor (CAR), i.e. it is a fusion of anti-CD19-41BB-CD3z, and only contains part of 41BB and CD3z. The CAR sequence is added to the Uniprot human database with 41BB and CD3z sequences. I also know that there is no native 41BB and CD3z expressed in the sample. 

If I try to draw it out as the following, (, unidentified peptides; |, identified peptides 
CD3z                        ((((((((((((((((((----------------------------------------------------------------------||||||||||||||||----|||||||----|||||----||||----||||||
41BB                                                     ((((((((((--------------------------------------|||||||||||||||
CAR                                                                      ((((((((((((((((((----||||||||-------|||||||||||||||-----||||||||||||||||----|||||||----|||||----||||----||||||

So there are 5 common peptides between CAR and CD3z, 1 common peptide between CAR and 41BB, 1 unique peptide for CAR, no unique peptide for CD3z and 41BB. In this case, I prefer Fragpipe approach in outputing only the CAR in the Genes column. Because there is no unique peptide for CD3z and I also know the truth. However, I also found Fragpipe approach problematic, in other cases. For example, it reports a single Histone H3, but there are many of them and are indistinguishable.

**Reply by WeiqiangChen** [reply]
I also find the default protein grouping very confusing in diaNN. If 3 proteins A, B and C, all have unique peptides identified and also some common peptides for ABC. DIANN will give protein groups, A, B, C and ABC, while MaxQuant, Spectronaut, FragPipe just give protein groups A, B and C.  Then diaNN will quantify 4 protein groups and others only 3. One protein showing up in multiple protein groups really does not make much sense. What should the end-users do with the much complex protein intensity table?

**Comment by hguturu** [comment]
What does DIA-NN do with `--reannotate` if the library was built with different parameters than the DIA-NN digest rules (e.g. the default workflow for MSFragger suggests peptides 7-50 and 2 miscleavages, while the default for DIANN is 7-30 is 1 miscleavage). The mass ranges are also different. Do the peptides in the library that dont map to the fasta due to different digest keep their old annotations? In that cause is the protein inference effected/incorrect? What is the best way to ensure that reannotate works consistently for the library? Do we have to set MSFragger to the same defaults as DIA-NN?

**Reply by vdemichev** [reply]
--reannotate works taking into account the enzyme specificity set in DIA-NN. Length & number of missed cleavages & mass ranges have no effect here.

**Reply by hguturu** [reply]
So as long as the same digest rule (e.g. K*,R*) is used in both MSFragger and DIA-NN, we will still be able to reannotate the library without knowing the parameters of the MSFragger? i.e. even if MSFragger outputs a spectral library for peptide of length 49 with 2 miscleavages, DIA-NN will still be able to assign that peptide to a fasta sequence even though by default it only digests up to peptide of length 30 with a max of 1 miscleavage?

Additionally, I noticed the workflow at https://www.nature.com/articles/s41467-022-31492-0#Sec7 turns on Oxidation of methionine, acetylation of protein N-termini, −18.0106 Da on N-terminal Glutamic acid, and −17.0265 Da on N-terminal Glutamine and Cysteine as variable modifications. Is it fine if those are in the library and we run DIA-NN without using --var-mod and --monitor-mod? Or does that risk high FDR and its better we build the library without any of these variable mods?

**Reply by vdemichev** [reply]
Yes. Also fine with modifications. --var-mod and --monitor-mod are needed to ensure the correct peptidoform identification if this is important for the specific experiment.
